power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -21,114 +25,242 @@ namespace detail = power_grid_model::optimizer::detail;
+                add_to_edge(edges, edge_props, to_node, from_node, {unregulated_idx, 1});
+            }
+        }
+    }
+}
+
+template <std::derived_from<Transformer> Component, class ComponentContainer>
+    requires main_core::model_component_state_c<main_core::MainModelState, ComponentContainer, Component>
+constexpr void add_edge(main_core::MainModelState<ComponentContainer> const& state,
+                        RegulatedObjects const& regulated_objects, TrafoGraphEdges& edges,
+                        TrafoGraphEdgeProperties& edge_props) {
+    for (auto const& transformer : state.components.template citer<Transformer>()) {
+        if (!transformer.from_status() || !transformer.to_status()) {
+            continue;
+        }
+        auto const& from_node = transformer.from_node();
+        auto const& to_node = transformer.to_node();
+
+        if (regulated_objects.transformers.contains(transformer.id())) {
+            auto const tap_at_from_side = transformer.tap_side() == BranchSide::from;
+            auto const& tap_side_node = tap_at_from_side ? from_node : to_node;
+            auto const& non_tap_side_node = tap_at_from_side ? to_node : from_node;
+            if (get_component<Node>(state, tap_side_node).u_rated() <
+                get_component<Node>(state, non_tap_side_node).u_rated()) {
+                throw AutomaticTapCalculationError(transformer.id());
+            }
+            add_to_edge(edges, edge_props, tap_side_node, non_tap_side_node,
+                        {main_core::get_component_idx_by_id(state, transformer.id()), 1});
+        } else {
+            add_to_edge(edges, edge_props, from_node, to_node, {unregulated_idx, 1});
+            add_to_edge(edges, edge_props, to_node, from_node, {unregulated_idx, 1});
+        }
+    }
+}
+
+template <typename Component>
+concept non_regulating_branch_c = std::derived_from<Link, Component> || std::derived_from<Line, Component>;
+
+template <non_regulating_branch_c Component, class ComponentContainer>
+    requires main_core::model_component_state_c<main_core::MainModelState, ComponentContainer, Component>
+constexpr void add_edge(main_core::MainModelState<ComponentContainer> const& state,
+                        RegulatedObjects const& /* regulated_objects */, TrafoGraphEdges& edges,
+                        TrafoGraphEdgeProperties& edge_props) {
+    auto const& iter = state.components.template citer<Component>();
+    edges.reserve(std::distance(iter.begin(), iter.end()) * 2);
+    edge_props.reserve(std::distance(iter.begin(), iter.end()) * 2);
+    for (auto const& branch : iter) {
+        if (!branch.from_status() || !branch.to_status()) {
+            continue;
+        }
+        add_to_edge(edges, edge_props, branch.from_node(), branch.to_node(), {unregulated_idx, 0});
+        add_to_edge(edges, edge_props, branch.to_node(), branch.from_node(), {unregulated_idx, 0});
+    }
+}
```

Comment:
This logic is identical for unregulated transformers. it's probably good to do something like this:

```suggestion
template <std::derived_from<Branch> Component, class ComponentContainer>
    requires main_core::model_component_state_c<main_core::MainModelState, ComponentContainer, Component>
constexpr void add_unregulated_edge(Component const& branch,
                        TrafoGraphEdges& edges,
                        TrafoGraphEdgeProperties& edge_props) {
    if (!branch.from_status() || !branch.to_status()) {
        continue;
    }
    add_to_edge(edges, edge_props, branch.from_node(), branch.to_node(), {unregulated_idx, 0});
    add_to_edge(edges, edge_props, branch.to_node(), branch.from_node(), {unregulated_idx, 0});
}

template <non_regulating_branch_c Component, class ComponentContainer>
    requires main_core::model_component_state_c<main_core::MainModelState, ComponentContainer, Component>
constexpr void add_edge(main_core::MainModelState<ComponentContainer> const& state,
                        RegulatedObjects const& regulated_objects, TrafoGraphEdges& edges,
                        TrafoGraphEdgeProperties& edge_props) {
    auto const& iter = state.components.template citer<Component>();
    edges.reserve(std::distance(iter.begin(), iter.end()) * 2);
    edge_props.reserve(std::distance(iter.begin(), iter.end()) * 2);
    for (auto const& branch : iter) {
        add_unregulated_edge<Component>(branch, edges, edge_props);
    }
}
```

Reply:
I thought of that, but its less desirable to refactor the few repeated lines with a new function that would need convoluted concept checks. Hardly any less duplication was seen with `add_regulated_edge` and `add_unregulated_edge`.

Reply:
riiiiiight... i forgot that unregulated transformers have a different edge weight....

i was going to write the following. However, `Transformers` act like branches but then not entirely so that doesn't apply here

> well this also concerns testability:
> * the check whether something is regulated or not depends on whether it satisfies the `transformer_c` concept.
> * the adding of an edge depends on the `Branch` type
> 
> constraints should be as generic that only one function is needed to be tested (here: something can be automatic tap regulated if and only if it is a `transformer_c` with an updatable `tap_pos`) and so specific that overloads are not accidentally added to other things.
> 
> duplication is one thing but the current implementation may actually make the logic flow less obvious because